src/firmware/linux/guest/types.rs

```diff
@@ -1 +1,388 @@
+    ///
+    /// | Bit(s) | Name          | Description                                                                                                    |
+    /// |--------|---------------|----------------------------------------------------------------------------------------------------------------|
+    /// | 7:0    | ABI_MINOR     | The minimum ABI minor version required for this guest to run.                                                  |
+    /// | 15:8   | ABI_MAJOR     | The minimum ABI major version required for this guest to run.                                                  |
+    /// | 16     | SMT           | 0: SMT is disallowed.<br>1: SMT is allowed.                                                               |
+    /// | 17     | -             | Reserved. Must be one.                                                                                         |
+    /// | 18     | MIGRATE_MA    | 0: Association with a migration agent is disallowed.<br>1: Association with a migration agent is allowed. |
+    /// | 19     | DEBUG         | 0: Debugging is disallowed.<br>1: Debuggin is allowed.                                                   |
+    /// | 20     | SINGLE_SOCKET | 0: Guest can be activated on multiple sockets.<br>1: Guest can only be activated on one socket.           |
+    ///
+    #[derive(Default)]
+    #[repr(C)]
+    pub struct SnpGuestPolicy(u64);
+    impl Debug;
+    get_api_minor, set_api_minor: 7, 0;
+    get_abi_major, set_abi_major: 15, 8;
+    get_smt, set_smt: 16, 16;
+    get_migrate_ma, set_migrate_ma: 18, 18;
+    get_debug, set_debug: 19, 19;
+    get_single_socket, set_single_socket: 20, 20;
+}
+
+/// The TCB_VERSION is a structure containing the security version numbers of each component in
+/// the trusted computing base (TCB) of the SNP firmware. A TCB_VERSION is associated with each
+/// image of firmware.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+
+pub struct SnpTcbVersion {
+    /// Current bootloader version. SVN of PSP Bootloader.
+    pub boot_loader: u8,
+    /// Current PSP OS version. SVN of PSP Operating System.
+    pub tee: u8,
+    reserved: [u8; 4],
+    /// Version of the SNP firmware. Security Version Number (SVN) of SNP firmware.
+    pub snp: u8,
+    /// Lowest current patch level of all the cores.
+    pub microcode: u8,
+}
+
+bitfield! {
+    /// A structure with a bit-field unsigned 64 bit integer:
+    /// Bit 0 representing the status of TSME enablement.
+    /// Bit 1 representing the status of SMT enablement.
+    /// Bits 2-63 are reserved.
+    #[derive(Default)]
+    #[repr(C)]
+    pub struct SnpPlatformInfo(u64);
+    impl Debug;
+    get_tsme_enabled, set_tsme_enabled: 0, 0;
+    get_smt_enabled, set_smt_enabled: 1, 1;
+}
+/// The default structure used for making requests to the PSP as a guest owner.
+pub struct SnpGuestRequest<'a, 'b, Req, Rsp> {
```

Comment:
I think this more belongs in `ioctl.rs`, as its used for each ioctl. This is similar to the `sev_issue_cmd` struct in the kernel, which is encapsulated by the `Command` struct.

Reply:
Will address in next commit.